### ISSUE TITLE
timedelta doesn't parse fractional seconds

### ISSUE THREAD

**Issue by brianmedigate** [pydantic/pydantic#8814]
### Initial Checks

- [X] I confirm that I'm using Pydantic V2

### Description

A field of the `timedelta` type, when given a `float`, seems to floor it to an `int` before creating a `timedelta`, even though `timedelta` supports fractional seconds.

### Example Code

```Python
from datetime import timedelta
from pydantic import BaseModel

class Foo(BaseModel):
    t: timedelta

print(repr(Foo(t=.5).t))
# datetime.timedelta(0)

print(repr(timedelta(seconds=.5)))
# datetime.timedelta(microseconds=500000)
```

### Python, Pydantic & OS Version

```Text
$ python -c "import pydantic.version; print(pydantic.version.version_info())"
             pydantic version: 2.6.1
        pydantic-core version: 2.16.2
          pydantic-core build: profile=release pgo=true
                 install path: .../venv/lib/python3.8/site-packages/pydantic
               python version: 3.8.18 (default, Aug 25 2023, 13:20:30)  [GCC 11.4.0]
                     platform: Linux-6.1.0-1033-oem-x86_64-with-glibc2.35
             related packages: email-validator-2.1.0.post1 typing_extensions-4.9.0 fastapi-0.109.0 pydantic-settings-2.1.0 mypy-1.8.0
                       commit: unknown
```
```

**Comment by sydney-runkle** [comment]
@brianmedigate,

When I run this on `main` with the latest version of `speedate`, I get the following:

```py
from datetime import timedelta
from pydantic import BaseModel

class Foo(BaseModel):
    t: timedelta

print(repr(Foo(t=.5).t))
# datetime.timedelta(microseconds=500000)

print(repr(timedelta(seconds=.5)))
# datetime.timedelta(microseconds=500000)
```

Additionally, with this extra config setting, I get the following behavior:

```py
from datetime import timedelta

from pydantic import BaseModel, ConfigDict

class Foo(BaseModel):
    t: timedelta

    model_config = ConfigDict(ser_json_timedelta='float')


print(repr(Foo(t=.5).t))
# datetime.timedelta(microseconds=500000)

print(repr(timedelta(seconds=.5)))
# datetime.timedelta(microseconds=500000)

print(Foo(t=.5).model_dump_json())
#> {"t":0.5}

print(Foo(t=timedelta(seconds=.5)).model_dump_json())
#> {"t":0.5}
```

Marking this as resolved, as I think this shows that the behavior you're looking for is present.

**Comment by brianmedigate** [comment]
great, thanks!

**Comment by timkpaine** [comment]
I'm still seeing this behavior on the latest release (2.6.3 from last week):

```python
  from datetime import timedelta
  from pydantic import BaseModel, __version__

class Foo(BaseModel):
    t: timedelta

print(repr(Foo(t=.5).t))
print(f"Pydantic: {__version__}")
```

```raw
datetime.timedelta(0)
Pydantic: 2.6.3
```

**Comment by brianmedigate** [comment]
same with 2.7.0. @sydney-runkle do you mind taking a look?

**Comment by timkpaine** [comment]
It would be good to reopen until it's available imo

**Comment by sydney-runkle** [comment]
Hey folks,

I'm getting the following with 2.7.1:

```py
from datetime import timedelta
from pydantic import BaseModel, __version__

class Foo(BaseModel):
    t: timedelta

print(repr(Foo(t=.5).t))
#> datetime.timedelta(microseconds=500000)
print(f"Pydantic: {__version__}")
#> Pydantic: 2.7.1
```

Apologies for the delayed response, the ping slipped through my email!

Lmk if you're still experiencing this issue, we can definitely reopen if it's still a problem!

**Comment by brianmedigate** [comment]
Yes, same behavior
```python
from datetime import timedelta
from pydantic import BaseModel, __version__

class Foo(BaseModel):
    t: timedelta

print(repr(Foo(t=.5).t))
#> datetime.timedelta(0)
print(f"Pydantic: {__version__}")
#> Pydantic: 2.7.1
```

**Comment by sydney-runkle** [comment]
@brianmedigate,

What version of `pydantic-core` are you using? Maybe we didn't bump the dependency correctly for `speedate`. If so, I can fix that in 2.7.2!

**Comment by brianmedigate** [comment]
2.18.2

**Comment by sydney-runkle** [comment]
@brianmedigate,

Can you share what version of `speedate` you're using? Working on our prep for 2.7.2 now and want to make sure everything is ironed out.

**Comment by brianmedigate** [comment]
How can I check? Isn't it a compile-time dependency of `pydantic_core`, so wouldn't it be implied by the version of `pydantic_core` that I'm using?

**Comment by sydney-runkle** [comment]
@davidhewitt, any advice here?

**Comment by davidhewitt** [comment]
I can reproduce with Pydantic 2.7.1; the important point is the Python version 3.8. I suspect also was broken on 3.9.

The good news is that it's already fixed in 2.7.2. I'm pretty sure this was a case of older versions suffering from integer promotion, similar to https://github.com/pydantic/pydantic/issues/9018, which was fixed in https://github.com/pydantic/pydantic-core/pull/1288.

**Comment by brianmedigate** [comment]
I can confirm that it's fixed for me in 2.7.2. Thanks!
